app/websocket/websocketclient.c

```diff
@@ -133,6 +127,37 @@ static void generateSecKeys(char **key, char **expectedKey) {
   os_free(keyEncrypted);
 }
 
+static int headers_length(header_t *headers) {
+  int length = 0;
+  for(; headers->key; headers++)
+    length += strlen(headers->key) + strlen(headers->value) + 4;
+  return length;
+}
+
+static int sprintf_headers(char *buf, ...) {
+  int len = 0;
+  va_list args;
+  va_start(args, buf);
+  for(header_t *header_set = va_arg(args, header_t *); header_set; header_set = va_arg(args, header_t *))
+    for(header_t *header = header_set; header->key; header++) {
+      va_list args2;
+      va_start(args2, buf);
+      for(header_t *header_set2 = va_arg(args2, header_t *); header_set2; header_set2 = va_arg(args2, header_t *))
+        for(header_t *header2 = header_set2; header2->key; header2++) {
+          if(header == header2)
+            goto ok;
+          if(!strcasecmp(header->key, header2->key))
+            goto skip;
+        }
+ok:
+      len += os_sprintf(buf + len, "%s: %s\r\n", header->key, header->value);
```

Comment:
Are you sure about high stack usage in above os_sprintf call? os_sprintf writes to provided buffer, formatting string is very simple, I cannot imagine how it can use a lot of stack memory. Can you explain / point some docs?